src/bp/nlu-core/entities/duckling-extractor/index.ts

```diff
@@ -0,0 +1,174 @@
+        }
+      }
+    } catch (err) {
+      console.error('could not load duckling cache')
+    }
+  }
+
+  public async extractMultiple(
+    inputs: string[],
+    lang: string,
+    useCache?: boolean
+  ): Promise<EntityExtractionResult[][]> {
+    if (!DucklingEntityExtractor.enabled) {
+      return Array(inputs.length).fill([])
+    }
+    const options = {
+      lang,
+      tz: this._getTz(),
+      refTime: Date.now()
+    }
+
+    const [cached, toFetch] = inputs.reduce(
+      ([cached, toFetch], input, idx) => {
+        if (useCache && DucklingEntityExtractor._cache.has(input)) {
+          const entities = DucklingEntityExtractor._cache.get(input)
+          return [[...cached, { input, idx, entities }], toFetch]
+        } else {
+          return [cached, [...toFetch, { input, idx }]]
+        }
+      },
+      [[], []] as [KeyedItem[], KeyedItem[]]
+    )
+
+    const chunks = _.chunk(toFetch, BATCH_SIZE)
+    const batchedRes = await Promise.mapSeries(chunks, c => this._extractBatch(c, options))
+
+    return _.chain(batchedRes)
+      .flatten()
+      .concat(cached)
+      .orderBy('idx')
+      .map(x => x.entities!)
+      .value()
+  }
+
+  public async extract(input: string, lang: string, useCache?: boolean): Promise<EntityExtractionResult[]> {
+    return (await this.extractMultiple([input], lang, useCache))[0]
+  }
+
+  private async _dumpCache() {
+    try {
+      await ensureFile(CACHE_PATH)
+      await writeJson(CACHE_PATH, DucklingEntityExtractor._cache.dump())
+    } catch (err) {
+      this.logger?.error(`could not persist system entities cache, error ${err.message}`, err)
+      this._cacheDumpEnabled = false
+    }
+  }
+
+  private _onCacheChanged = _.debounce(async () => {
+    if (this._cacheDumpEnabled) {
+      await this._dumpCache()
+    }
+  }, ms('10s'))
+
+  private async _extractBatch(batch: KeyedItem[], params: DucklingParams): Promise<KeyedItem[]> {
+    if (_.isEmpty(batch)) {
+      return []
+    }
+    // trailing JOIN_CHAR so we have n joints and n examples
+    const strBatch = batch.map(x => x.input)
+    const concatBatch = strBatch.join(JOIN_CHAR) + JOIN_CHAR
+    const batchEntities = await this._fetchDuckling(concatBatch, params)
+    const splitLocations = extractPattern(concatBatch, new RegExp(JOIN_CHAR)).map(v => v.sourceIndex)
+    const entities = splitLocations.map((to, idx, locs) => {
+      const from = idx === 0 ? 0 : locs[idx - 1] + JOIN_CHAR.length
+      return batchEntities
+        .filter(e => e.start >= from && e.end <= to)
+        .map(e => ({
+          ...e,
+          start: e.start - from,
+          end: e.end - from
+        }))
+    })
+    await this._cacheBatchResults(strBatch, entities)
+
+    return batch.map((batchItm, i) => ({ ...batchItm, entities: entities[i] }))
+  }
+
+  private async _fetchDuckling(text: string, params: DucklingParams): Promise<EntityExtractionResult[]> {
+    const duckReturn = await this._provider.fetchDuckling(text, params)
```

Comment:
this has changed (extracted initialisation of duckling inside duckling provider)